drivers/mmc/card/emmc_mfgdata.c

```diff
@@ -0,0 +1,328 @@
+	int i = 0;
+
+	pos = 0;
+	if (!parse_header()) {
+		pr_info("mfgdata: no header found\n");
+		return false;
+	}
+
+	/* Pass 1: count number of entries */
+	num_entries = 0;
+	tmp = pos;
+	while ((r = parse_entry(&tag, &length, &value)) == 1)
+		num_entries++;
+
+	if (r == -1) {
+		pr_info("mfgdata: parse failure at %d\n", pos);
+		num_entries = 0;
+		return false;
+	}
+
+	mfgdata_entries = kcalloc(num_entries, sizeof(struct mfgdata_entry),
+				  GFP_KERNEL);
+	pos = tmp;
+
+	/* Pass 2: parse entries, create sysfs attributes */
+	for (i = 0; i < num_entries; i++) {
+		struct mfgdata_entry *entry = &mfgdata_entries[i];
+		r = parse_entry(&entry->tag, &entry->length,
+				&entry->data);
+		if (r != 1) {
+			pr_info("mfgdata: parse failure on tag %d\n", i);
+			goto err;
+		}
+
+		validate_entry(entry);
+		if (!entry->valid)
+			continue;
+
+		sysfs_attr_init(&entry->attr->attr);
+		entry->attr.attr.name = entry->tag;
+		entry->attr.attr.mode = 0400;
+		entry->attr.show = mfgdata_attr_show;
+		if (device_create_file(mfgdata_dev, &entry->attr)) {
+			pr_warn("mfgdata: failed creating attr for %s\n", tag);
+			goto err;
+		}
+	}
+
+	return true;
+
+err:
+	num_entries = 0;
+	kfree(mfgdata_entries);
+	mfgdata_entries = NULL;
+	return false;
+}
+
+static void prepare_mrq(struct mmc_card *card,
+			struct mmc_request *mrq, struct scatterlist *sg,
+			unsigned sg_len, u32 dev_addr, unsigned blocks,
+			unsigned blksz)
+{
+	BUG_ON(!mrq || !mrq->cmd || !mrq->data || !mrq->stop);
+
+	if (blocks > 1)
+		mrq->cmd->opcode = MMC_READ_MULTIPLE_BLOCK;
+	else
+		mrq->cmd->opcode = MMC_READ_SINGLE_BLOCK;
+
+	mrq->cmd->arg = dev_addr;
+	if (!mmc_card_blockaddr(card))
+		mrq->cmd->arg <<= 9;
+
+	mrq->cmd->flags = MMC_RSP_R1 | MMC_CMD_ADTC;
+
+	if (blocks == 1)
+		mrq->stop = NULL;
+	else {
+		mrq->stop->opcode = MMC_STOP_TRANSMISSION;
+		mrq->stop->arg = 0;
+		mrq->stop->flags = MMC_RSP_R1B | MMC_CMD_AC;
+	}
+
+	mrq->data->blksz = blksz;
+	mrq->data->blocks = blocks;
+	mrq->data->flags = MMC_DATA_READ;
+	mrq->data->sg = sg;
+	mrq->data->sg_len = sg_len;
+
+	mmc_set_data_timeout(mrq->data, card);
+}
+
+int emmc_mfgdata_parse(struct mmc_card *card, unsigned long capacity)
+{
+	unsigned int blocks = DATA_BLOB_SIZE >> card->csd.read_blkbits;
+	u32 addr = (capacity - 8192) >> card->csd.read_blkbits;
+	struct scatterlist sg;
+	struct mmc_request mrq = {0};
+	struct mmc_command cmd = {0};
+	struct mmc_command stop = {0};
+	struct mmc_data data = {0};
+
+	sg_init_one(&sg, mfgdata_buf, DATA_BLOB_SIZE);
+	mrq.cmd = &cmd;
+	mrq.data = &data;
+	mrq.stop = &stop;
+	prepare_mrq(card, &mrq, &sg, 1, addr, blocks,
+		    1 << card->csd.read_blkbits);
+	mmc_wait_for_req(card->host, &mrq);
+
+	if (cmd.error) {
+		pr_err("mfgdata: command error %d\n", cmd.error);
+		return cmd.error;
+	}
+
+	if (data.error) {
+		pr_err("mfgdata: data error %d\n", cmd.error);
+		return data.error;
+	}
+
+	parse();
+	pr_info("mfgdata: read %d entries\n", num_entries);
+	mfgdata_initialized = true;
```

Comment:
Yes. It's so that drivers can distinguish between "no data was returned because we haven't yet found the eMMC device, please try again later" vs "no data was returned because it's not there".